flips/20220715-cadence-purity-analysis.md

```diff
@@ -0,0 +1,211 @@
+# Function Purity 
+
+| Status        | (Proposed)       |
+:-------------- |:---------------------------------------------------- |
+| **FLIP #**    | [1056](https://github.com/onflow/flow/pull/1056)     |
+| **Author(s)** | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Sponsor**   | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Updated**   | 2022-08-25                                           |
+
+## Objective
+
+This FLIP proposes to add a new syntax and accompanying semantic analysis to determine or 
+enforce that a given function or method is "view", that is, lacking in side effects. We would
+then add an additional checker rule to enforce purity in function conditions. 
+
+## Motivation and User Benefit
+
+An [issue](https://github.com/onflow/cadence/issues/1805) with function pre- and post-conditions allows arbitrary functions to be
+called in these condition blocks, which can allow implementers to write underhanded code by modifying state inside of a condition. 
+By allowing users to annotate their functions as `view`, we can enforce that these functions have no side effects that modify state
+and thus permit these functions (and only these functions) to be called from within a condition. 
+
+## Design Proposal
+
+### Parsing
+
+This adds a new keyword to the langauge: `view`. This is optionally placed before the `fun` keyword, but after the access modifier, 
+in function declarations and expressions. So, for example, all of the following are valid:
+
+```cadence
+pub view fun foo(): Void {}
+```
+
+```cadence
+view fun foo(): Void {}
+```
+
+```cadence
+let x = view fun(): Void {}
+```
+
+```cadence
+pub struct S {
+    pub view fun foo(): Void {}
+    view init()
+}
+```
+
+```cadence
+pub resource interface R {
+    priv view fun foo(): Void {}
```

Comment:
Just as a side note for the implementation: The type checker will need to ensure that the function in the concrete type that implements the interface has the modifier it is given for the function in the interface. This conformance has to be checked for interfaces and nested type requirements

Reply:
Yes; the implementation already does this!